ACPs/77-reinventing-subnets/README.md

```diff
@@ -257,6 +280,8 @@ Issuing a `SetSubnetValidatorWeightTx` with `Weight` of `0` will remove the Subn
 
 All state related to the Subnet Validator being removed will be removed from the P-Chain's active state (BLS key, NodeID etc). This validator can be newly added to the Subnet's validator set using the `RegisterSubnetValidatorTx` flow.
 
+Since altering a Subnet's validator set after a `ConvertSubnetTx` requires a valid Warp message from the Subnet's current validator set, it is explicitly disallowed for a Subnet to remove it's only validator, which would render the Subnet unable to produce any further valid Warp messages. If a Subnet only has a single validator, it is invalid to set that validator's weight to 0 via a `SetSubnetValidatorWeightTx`.
+
```

Comment:
If the Subnet only has a single validator, that validator is explicitly signing that they want to be removed. I wonder if it would be a better design to garbage collect Subnets that have 0 validators than somewhat arbitrarily require that the last Validator to be removed is the only revival mechanism. If the last Validator stops paying the continuous fee, the Subnet is effectively dead. 

Relatedly, if 50% of a Subnet's weight stops paying the continuous fee, the Subnet is effectively dead. I believe the guidance there is to fork the Subnet with those Validators removed. Would that be a suitable answer for Subnets that get in this position?

Reply:
Would defer to @StephenButtolph here.

But IIUC, validators stopping paying the continuous fee or directly disabling themselves on the P-Chain is functionally different from the final validators removing themselves because they could always re-up/re-enable themselves to revive the Subnet in that case, where if they were fully removed then there is zero recovery path.

Reply:
- I think it might make sense to move the "If the last validator leaves then the subnet is deleted" to a followup... It would fit very easily into a future change if we were to just disallow it in this ACP.
- I don't think that the subnet should be removed if too much of the stake becomes deactivated (As the subnet isn't known to be dead, because balances can be increased without a Warp message).

Reply:
I agree that Subnets shouldn't be deleted if too much of the stake becomes deactivated. For explicit removals via `SetSubnetValidatorWeightTx` with `Weight=0` is a different story. 

Disallowing the last one to be removed seems like an odd design pattern to me since the validator is explicitly signing off on its own removal. Shouldn't it be allowed that they are able to remove themselves? 

The `RecoverSubnetTx` was designed to guard against this since we could kick-back recovery path to the `Owner`. Instead of re-introducing that tx (which had other issues as @michaelkaplan13 noted), I think it's reasonable to say the Subnet can no longer be revived. If anybody wants to revive the Subnet, you'll need to fork the state and create a new Subnet.

Reply:
I think this is mostly a UX question for this extreme edge case.

In the event that the only Subnet validator wants to kill the Subnet forever, it seems like it would be functionally equivalent for them to just stop/delete their node for that Subnet as compared to setting their weight to 0 on the P-Chain.

On the other hand though, explicitly disallowing the destructive action on the P-chain could act as a defensive-in-depth guard rail for a single validator accidentally signing to remove themselves when they don't actually want to kill the Subnet forever.

I would prefer to play it on the safer side of not allowing the more destructive action for now (which would be more likely to occur accidentally immediately following activation as developers are figuring out how to properly integrate), and if the number of effectively "dead" Subnets in the future ever becomes an issue, then it could be changes with a separate ACP.

Reply:
Having a guard rail for this makes sense to me. However, I'm not sure the P-Chain itself is the right place for it. Forcing a validator to stay in the validator set when they're explicitly signing that they want to be removed still seems like a strange design with the only benefit being this guard rail. While signing for removal is functionally equivalent to simply stopping/deleting their node, it doesn't remove them from the validator set meaning their NodeID is still associated with the Subnet. This could potentially cause regulatory concerns as well (iirc some Primary Network Validators were concerned about not being able to early eject before the EndTime)

Should this be handled on the Subnet Manager contract side instead? If you want to remove the last validator of the set, you have to remove it through a different method than the standard method. I guess this could be viewed kind of like a "Are you sure you want to delete this Subnet?" prompt. That should avoid the accidental signature concern you raised

Reply:
To play devil's advocate here, what's the gained benefit from not explicitly enforcing it on the P-chain when it's functionally equivalent? I don't think there are any regulatory concerns regarding a completely halted chain that will never produce any more blocks or activity, and node IDs are freely rotated. Validator manager contract wise, the final validator actually would never be able to completely remove themselves because the process requires posting an ack of the removal on the P-chain back on the L1, which would at that point not be able to produce a new block.

To me it seems like removals of the final validator are far more likely to be an accident than intentional, and even when intentional the benefit/motivation to do so is not very clear.

Reply:
The main benefit is a validator isn't locked into the Subnet's validator set on the P-chain. A validator is able to willingly turn off or delete their node. With this guardrail, the P-chain would prohibit them from removing themselves.

> Validator manager contract wise, the final validator actually would never be able to completely remove themselves because the process requires posting an ack of the removal on the P-chain back on the L1, which would at that point not be able to produce a new block.

The canonical validator set of a Subnet lives on the P-chain. If a validator is removed on the P-chain, they are removed from the Subnet's canonical validator set and will no longer participate in consensus. Sending the ack back to the L1 isn't required to stop participating.

> To me it seems like removals of the final validator are far more likely to be an accident than intentional, and even when intentional the benefit/motivation to do so is not very clear.

I think my earlier suggestion of requiring removal through a different method on the contract would sufficiently prevent accidental removals without locking in any validators on the P-chain.

> I don't think there are any regulatory concerns regarding a completely halted chain that will never produce any more blocks or activity, and node IDs are freely rotated.

This isn't true, otherwise this wouldn't be a guardrail. The validator can come back online via the `IncreaseBalanceTx` which would revive the Subnet. It's not guaranteed that the validator will never participate in the Subnet again, potentially causing regulatory concerns. 

Fwiw I don't think the regulatory stuff is that big of a point. I was just reminded of some concerns Primary Network Validators brought up when thinking about this guardrail. I'm mainly disagreeing with placing a somewhat arbitrary restriction in the P-chain that isn't required for safety/liveness of the P-chain. 

This feels to me like a Subnet-level guardrail, not a Primary Network-level guardrail. For example, a Subnet may want two PAYG validators always attached to the Subnet instead of just one. If they want to do that, that would need to live in the Subnet contract.

Reply:
Yeah, fair enough. I was just talking with @StephenButtolph about this topic, his take on it is slightly different:
- If the only Subnet validator is removed on the P-Chain (permanently killing the Subnet), then all of the other Subnet state should also be cleaned up and removed from the P-Chain.
- How to properly remove the entire state of a Subnet is outside the scope of the ACP-77 and the Etna upgrade.
- Allowing the only Subnet validator to be removed should be deferred to a later ACP, which includes the full Subnet state clean up. Until then, permanently killing Subnets without removing them from state will be accumulating tech debt.

Reply:
As far as I understand, cleaning up and removing dead Subnets from the P-chain state would not constitute an execution level change. This would mean it is not required to scope out how to properly remove a dead Subnet from the P-chain state in ACP-77, the initial Etna release, or a future ACP. ACP-77 could go forward without having any restrictions on the removal of the last PAYG validator. AvalancheGo can resolve the issue of cleaning up a dead Subnet at any later point in time. This tech debt feels like a minor problem since `CreateSubnetTx` can currently be issued without any cleanup of a Subnet's state, even when the Subnet is functionally dead (e.g. https://subnets.avax.network/subnets/2TSYtJmLdmBJCTcEP6ZeQ3d7urg1GbrKtVMMdd7nmz46z22PSC).

If we include this restriction in Etna, the minimum point at which we can enable Subnet validators to remove themselves if they are the last validator is in the F upgrade. On the other hand, we can not have this restriction in Etna and clean up the state in a minor release of AvalancheGo.

Given the above, I think we should not introduce this restriction in Etna. Would love to hear @StephenButtolph's thoughts on this.

Reply:
You're right, cleaning up the subnet state wouldn't need to be a consensus decision.

- I feel like allowing deletion of the last SoV _could_ brick a subnet.
- I feel like there is no clear reason to allow the last SoV to leave the subnet.

Specifically:
> The main benefit is a validator isn't locked into the Subnet's validator set on the P-chain. A validator is able to willingly turn off or delete their node. With this guardrail, the P-chain would prohibit them from removing themselves.

All the validator needs to do to stop validating a subnet, is remove the subnetID from their `--track-subnets` flag. In fact, it's possible to add any validator to any subnet. Additionally, doing so could be done with a BLS key that the actual owner of the nodeID doesn't have access to. So if we actually care about this case, we need to restructure the BLS PoP.

All-in-all, I feel like just disallowing this behavior is the most future-proof thing to do. We can circle back to how we want to handle this edge case in the future.

Reply:
> In fact, it's possible to add any validator to any subnet. Doing so can be done with a BLS key that the actual owner of the nodeID doesn't have access to. So, if we actually care about this case, we'd need to restructure the BLS PoP.

The BLS PoP point is a good one. A subtle difference here is that we'd be restricting removal even when the NodeID is signing off on it by virtue of being the last validator in the Subnet.

I've thought about it more and re-read the discussion here. I definitely agree that allowing deletion of the last SoV could brick a Subnet. Forking and re-creating a new Subnet at this point is sub-optimal to say the least (especially without any tooling). There are strong merits (as both @StephenButtolph and @michaelkaplan13 pointed out) on having the guardrail live in the P-chain until at least F. Personally, I think the end-vision of this design would be removing this restriction (in a follow-up ACP), cleaning up the Subnet state entirely, and moving the guardrail to the Subnet Manager contract level.

Would it make sense to include some language to add color here? What do you guys think about:

```suggestion
If all Subnet Validators are removed, there are no valid Warp messages that can be produced to register new Subnet Validators through `RegisterSubnetValidatorTx`. With no validators, block production will halt and the Subnet is unrecoverable. To serve as a guardrail against this situation, a Subnet Validator removal will be considered invalid if the validator being removed is the last one. A future ACP can remove this guardrail as users get more familiar with the new Subnet mechanics and tooling matures to fork a Subnet.
```

Reply:
Just a side note here, one way for a Subnet to protect against death by empty validator set is to set its manager address on another chain, e.g. the C-Chain. This is the design we have chosen for Suzaku, where all contracts will be hosted on the C-Chain at it is the more decentralized and secure of the ecosystem today.

Reply:
The same problem applies. With an empty validator set, block production will halt. The only way to make progress is have a non-empty validator set which isn't guarded against by not using the Subnet for the manager contract. The only way to guard against it is to either prohibit an empty validator set on the P-Chain or on the manager contract.

Reply:
I think @Nuttymoon's point is that even if it is guarded against in the contracts (which I agree it should be), if the validator manager is on the L1 with a single validator itself, that validator could sign the arbitrary message to remove themselves even if the contract doesn't allow for emitting that message on-chain.

But if it's on the C-Chain and the contract doesn't allow emitting that message, then the single validator can't remove themselves in any way.

Reply:
Good point, I glossed over that when I first ready @Nuttymoon's comment. Putting it on the C-Chain does guard against the Subnet's validator set side-stepping the rules of the manager contract.